### PR TITLE
[@types/mongoose] Add orFail method

### PR DESCRIPTION
**Pull request by kesselb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://mongoosejs.com/docs/api.html#query_Query-orFail
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/mongoose/index.d.ts [hunk 1]
```diff
@@ -2021,6 +2021,30 @@ declare module "mongoose" {
...
+     *     // Throws "No docs found!" error if no docs match `{ foo: 'bar' }`
+     *     await Model.find({ foo: 'bar' }).orFail(new Error('No docs found!'));
+     *
+     *     // Throws "Not found" error if no document was found
+     *     await Model.findOneAndUpdate({ foo: 'bar' }, { name: 'test' }).
+     *       orFail(() => Error('Not found'));
+     *
+     * @param err optional error to throw if no docs match `filter`
+     */
+    orFail(err?: any): Query<any> & QueryHelpers;
```
Comment (Alorel) [inline types/mongoose/index.d.ts]: The parameter should be a function or error object.
```suggestion
    orFail(err?: Error | (() => Error)): Query<any> & QueryHelpers;
```

**Review by Alorel — CHANGES_REQUESTED** [review]

**Review by Alorel — APPROVED** [review]

**Comment by danielkesselberg** [comment]
I could need some advice about the return type :confused: In `mongoose/lib/query.js` most of the time `{Query} this` is used as return type. I'm not sure when to use `Query<any> & QueryHelpers` or `this`? Thanks!

**Comment by danielkesselberg** [comment]
I could need some advice about the return type :confused: In `mongoose/lib/query.js` most of the time `{Query} this` is used as return type. I'm not sure when to use `Query<any> & QueryHelpers` or `this`? Thanks!

**Comment by Alorel** [comment]
> I could need some advice about the return type  In `mongoose/lib/query.js` most of the time `{Query} this` is used as return type. I'm not sure when to use `Query<any> & QueryHelpers` or `this`? Thanks!

If the JS version returns `this`, the type definition should also return `this`

**Comment by Alorel** [comment]
> I could need some advice about the return type  In `mongoose/lib/query.js` most of the time `{Query} this` is used as return type. I'm not sure when to use `Query<any> & QueryHelpers` or `this`? Thanks!

If the JS version returns `this`, the type definition should also return `this`

**Comment by danielkesselberg** [comment]
Thank you!

**Pull request merged**